### ISSUE TITLE
NeurIPS '19 | Generative modeling by estimating gradients of the data distribution.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#25]
Song Y, Ermon S. [Generative modeling by estimating gradients of the data distribution](https://arxiv.org/pdf/1907.05600.pdf).

**Comment by NorbertZheng** [comment]
## Related Reference
 - Sander Dieleman. [Diffusion models are autoencoders](https://benanne.github.io/2022/01/31/diffusion.html).

**Comment by NorbertZheng** [comment]
## Introduction
Diffusion models took off like a rocket at the end of 2019, after the publication of Song & Ermon’s [seminal paper](https://arxiv.org/abs/1907.05600). In this paper-reading, I highlight a connection to another type of model: the venerable autoencoder.

**Comment by NorbertZheng** [comment]
## Diffusion Models
![diffuse2](https://user-images.githubusercontent.com/37235023/170860949-897bd4ce-e283-4c5f-879a-6b0cc3f6d3bf.jpg)
Diffusion models are fast becoming the go-to model for any task that requires producing perceptual signals, such as images and sound. They provide similar fidelity as alternatives based on generative adversarial nets (GANs) #22 or autoregressive models, but with
 - much better mode converge than the GANs.
 - and a faster and more flexible sampling procedure compared to the autoregressive models.

In a nutshell, diffusion models are constructed by first describing a procedure for gradually turning data into noise, and then training a neural network that learns to invert this procedure step-by-step. 
 - **Each of these steps consists of taking a noisy input and making it slightly less noisy, by filling in some of the information obscured by the noise.**
 
If you start from pure noise and do this enough times, it turns out you can generate data this way!

Diffusion models have been around for a while, but really took off at the end of 2019. The ideas are young enough that the field hasn't really settled one particular convention or paradigm to describe them, which means almost every paper uses a slightly different framing, and often a different notation as well. This can make it quite challenging to see the bigger picture when trawling through the literature, of which there is already a lot! Diffusion models go by many names:
 - denoising diffusion probabilistic models (DDPMs).
 - score-based generative models.
 - generative diffusion processes.

Some people just call them **energy-based models (EBMs)**, of which they technically are a special case.

My personal favorite perspective starts from the idea of score matching, [Hyvarinen et. al.](https://www.jmlr.org/papers/v6/hyvarinen05a.html), and uses a formalism based on stochastic differential equations (SDEs), [Sohl-Dickstein et. al.](https://arxiv.org/abs/2011.13456). For an in-depth treatment of diffusion models from this perspective, I strongly recommend [Yang Song's richly illustrated blog post](https://yang-song.github.io/blog/2021/score/) (which also comes with code and colabs). It is especially enlightening with regards to the connection between all these different perspectives. If you are familiar with variational autoencoders, you may find [Lilian Weng](https://lilianweng.github.io/lil-log/2021/07/11/diffusion-models.html) or [Jakub Tomczak](https://jmtomczak.github.io/blog/10/10_ddgms_lvm_p2.html)'s takes on this model family more approachable.

**Comment by NorbertZheng** [comment]
## Denoising Autoencoders
![bottleneck](https://user-images.githubusercontent.com/37235023/170862196-b28cf11e-16f4-45a2-9c99-248c7723d544.jpg)
Autoencoders are neural networks that are trained to predict their input. In and of itself, this is a trivial and meaningless task, but it becomes more interesting
 - when the network architecture is restricted in some way.
 - when the input is corrupted and the network has to learn to undo this corruption.

A typical architectural restriction is to introduce some sort of bottleneck, which limits the amount of information that can pass through. This implies that the network must learn to encode the most important information efficiently to be able to pass it through the bottleneck, in order to be able to accurately reconstruct the input. Such a bottleneck can be created by reducing the capacity of a particular layer of the network,
 - by introducing quantisation (as in VQ-VAEs).
 - or by applying some form of regularization to it during training (as in VAEs or contractive autoencoders).

The internal representation used in this bottleneck (often referred to as the latent representation) is what we are really after. It should capture the essence of the input, while discarding a lot of irrelevant detail.

Corrupting the input is another viable strategy to make autoencoders learn useful representations. One could argue that models with corrupted input are not autoencoders in the strictest sense, because the input and target output differ, but this is really a semantic discussion - one could just as well consider the corruption procedure part of the model itself, In practice, such models are typically referred to as denoising autoencoders.

Denoising autoencoders were actually some of the first true “deep learning” models: back when we hadn’t yet figured out how to reliably train neural networks deeper than a few layers with simple gradient descent, the prevalent approach was to pre-train networks layer by layer, and denoising autoencoders were frequently used for this purpose [Vincent et. al.](https://www.jmlr.org/papers/v11/vincent10a.html) (especially by Yoshua Bengio and colleagues at MILA – restricted Boltzmann machines were another option, favoured by Geoffrey Hinton and colleagues).

**Comment by NorbertZheng** [comment]
## One and the same?
![spiderman](https://user-images.githubusercontent.com/37235023/170864574-51a3a4a6-23a4-4941-991f-53deb13ff161.jpg)

**Comment by NorbertZheng** [comment]
## A tenuous connection?
![bridge](https://user-images.githubusercontent.com/37235023/171011384-6e670efd-2c97-44d1-bf06-d18fbc681e8c.jpg)
Of course, the title of this blog post is intentionally a bit facetious: while there is a deeper connection between diffusion models and autoencoders than many people realize, the models have completely different purposes and so are interchangeable.

There are two key differences with the denoising autoencoders of yore:
 - the additional input $t$ makes one single model able to handle many different noise levels with a single set of shared parameters.
 - we care about the output of the model, not the internal latent representation, so there is no need for a bottleneck. In fact, it would probably do more harm than good.

**Comment by NorbertZheng** [comment]
## Noise and Scale
![noisy_mountains](https://user-images.githubusercontent.com/37235023/171083013-b47fae19-67fb-4de1-9cda-656af6ae31a6.jpg)
I believe the idea of training a single model to handle many different noise levels with shared parameters is ultimately the key ingredient that made diffusion models really take off. [Song & Ermon](https://arxiv.org/abs/1907.05600) called them noise-conditional score networks (NCSNs) and provide a very lucid explanation of why this is important, which I won't repeat here.

The idea of using different noise levels in a single denoising autoencoder had previously been explored for representation learning, but not for generative modelling. Several works suggest gradually decreasing the level of noise over the course of training to improve the learnt representations, [Geras et. al.](https://arxiv.org/abs/1406.3269), [Chandra et. al.](https://link.springer.com/chapter/10.1007/978-3-319-12637-1_67), [Zhang et. al.](https://dl.acm.org/doi/abs/10.1007/s11704-016-6107-0). [Composite denoising autoencoders](https://link.springer.com/chapter/10.1007/978-3-319-46128-1_43) have multiple subnetworks that handle different noise levels, which is a step closer to the score networks that we use in diffusion models, though still missing the parameter sharing.

**Comment by NorbertZheng** [comment]
## Closing thoughts
![sunset](https://user-images.githubusercontent.com/37235023/171087160-a3d30a10-7f24-46b9-8286-d9e88ad7d28c.jpg)
So there you have it: diffusion models are autoencoders.  Sort of. When you squint a bit. Here are some key takeaways, to wrap up:
 - Learning to predict the score function $\nabla_{x}logp(x)$ of a distribution can be achieved by learning to denoise examples of that distribution. This is a core underlying idea that powers modern diffusion models.
 - Compared to denoising autoencoders, score networks in diffusion models can handle all noise levels with a single set of parameters, and do not have bottlenecks. But other than that, they do the same thing.
 - Noise levels and features scales are closely linked: high noise levels lead to models capturing large-scale features, low noise levels lead to models focusing on fine-grained features.
